src/Lykke.Service.LoggingAdapter.Services/HealthNotification/HealthNotifierStorage.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.Common.Log;
+using Lykke.Logs;
+using Lykke.Service.LoggingAdapter.Core.Domain.HealthNotification;
+using Lykke.SlackNotifications;
+
+namespace Lykke.Service.LoggingAdapter.Services.HealthNotification
+{
+    public class HealthNotifierStorage: IHealthNotifierStorage, IDisposable
+    {
+        private readonly ConcurrentDictionary<HealthNotifierBuilderSettings, IHealthNotifier> _healthNotifiers;
+        private readonly ILog _log;
+        private readonly ISlackNotificationsSender _slackNotificationsSender;
+
+        public HealthNotifierStorage(ILogFactory logFactory, ISlackNotificationsSender slackNotificationsSender)
+        {
+            _slackNotificationsSender = slackNotificationsSender;
+            _healthNotifiers = new ConcurrentDictionary<HealthNotifierBuilderSettings, IHealthNotifier>();
+            _log = logFactory.CreateLog(this);
+        }
+
+        public IHealthNotifier GetOrCreateHealthNotifier(string appName, string appVersion, string envInfo, ILogFactory logFactory)
+        {
+            var builderSettings = new HealthNotifierBuilderSettings(appName, appVersion, envInfo);
+
+            return _healthNotifiers.GetOrAdd(builderSettings, p =>
+            {
+                _log.Info($"Registering health notifier: appName {p.AppName}, appVersion {p.AppVersion}, envInfo {p.EnvInfo}");
```

Comment:
Note that in ConcurrentDictionary, this delegate could be executed multiple times for the same key, while key is not added to the dictionary yet.  Seems that OnDemandDataCache is more suitable here, instead of ConcurrentDictionary